### PR TITLE
[Transaction] Transaction buffer clear abort transactions

### PR DESCRIPTION
**Pull request by congbobo184**
## Motivation
Transaction buffer clear abort transactions.

## implement
check the aborts position's ledgerId whether exist in topic ```ManagedLedgerImpl```  ledgers.
### Verifying this change
Add the tests for it

Does this pull request potentially affect one of the following parts:
If yes was chosen, please highlight the changes

Dependencies (does it add or upgrade a dependency): (no)
The public API: (no)
The schema: (no)
The default values of configurations: (no)
The wire protocol: (no)
The rest endpoints: (no)
The admin cli options: (no)
Anything that affects deployment: (no)

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 1]
```diff
@@ -322,8 +322,12 @@ private void takeSnapshot() {
         });
     }
 
-
-
+    private void clearAbortTransaction() {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:325]: What about 'clearAbortedTransactions' ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 2]
```diff
@@ -322,8 +322,12 @@ private void takeSnapshot() {
...
+        while (!aborts.isEmpty() && !((ManagedLedgerImpl) topic.getManagedLedger())
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:326]: IIRC there is an ongoing effort to not use ManafedLedgerImpl but only the interface.
Is it necessary to do this cast?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 3]
```diff
@@ -322,8 +322,12 @@ private void takeSnapshot() {
...
+                .ledgerExists(aborts.get(aborts.firstKey()).getLedgerId())) {
+            aborts.remove(aborts.firstKey());
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:328]: Should be log the txid of the transactions that we are clearing?

**Review by eolivelli — COMMENTED** [review]
Overall looks good to me.

I left a few questions, can you please take a look?

**Review by sijie — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java (hunk 1 above)
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:325]: good suggestion

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java (hunk 3 above)
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:328]: good, i will add the debug level log for clear aborted transaction.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java (hunk 2 above)
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:326]: ```ledgerExists``` it not the ```ManagedLedger``` interface method.

**Review by eolivelli — APPROVED** [review]
LGTM

**Pull request merged**